### ISSUE TITLE
Flux Outpainting

### ISSUE THREAD

**Issue by fakerybakery** [huggingface/diffusers#9188]
**Is your feature request related to a problem? Please describe.**
Outpainting doesn't seem to be supported in Flux

**Describe the solution you'd like.**
Add outpainting to Flux

**Describe alternatives you've considered.**
N/A

**Additional context.**
#9135 introduces Flux inpainting

**Comment by asomoza** [comment]
Hi, we don't have any outpainting pipeline for any model, this is because outpainting can be done with the same inpainting pipeline you're referencing.

Outpainting is just inpainting but using just a part of the reference image to expand it.

**Comment by yiyixuxu** [comment]
closing this issue now 
we may publish an outputting guide with flux in the future when we have more tools available & when @asomoza has good ideas!

**Comment by asomoza** [comment]
Hi, I don't think you're doing anything wrong but currently this is a limitation of the model, you can search for the issues with flux and inpainting.

This is basically because:

- Flux doesn't have an inpainting model.
- Flux is distilled and really aesthetic (so it's hard that it does smooth transitions).
- It has a really low range of the strength you can use and get some decent results (which is really bad for inpainting).

Also you don't get that much benefit with a blurred mask with the inpainting mask since the pipeline transforms it to a binary mask.

There's still some couple of things you can try:

- Use a undistilled model but you'll need to modify the pipeline so you can use CFG.
- Use the [inpainting controlnet](https://huggingface.co/alimama-creative/FLUX.1-dev-Controlnet-Inpainting-Beta) but it will probably change the base image too.
- Use differential diffusion with a soft/blurred mask, but the pipeline it's still a [PR](https://github.com/huggingface/diffusers/pull/9268).

**Comment by Clement-Lelievre** [comment]
> Hi, I don't think you're doing anything wrong but currently this is a limitation of the model, you can search for the issues with flux and inpainting.
> 
> This is basically because:
> 
>     * Flux doesn't have an inpainting model.
> 
>     * Flux is distilled and really aesthetic (so it's hard that it does smooth transitions).
> 
>     * It has a really low range of the strength you can use and get some decent results (which is really bad for inpainting).
> 
> 
> Also you don't get that much benefit with a blurred mask with the inpainting mask since the pipeline transforms it to a binary mask.
> 
> There are still some couple of things you can try:
> 
>     * Use a undistilled model but you'll need to modify the pipeline so you can use CFG.
> 
>     * Use the [inpainting controlnet](https://huggingface.co/alimama-creative/FLUX.1-dev-Controlnet-Inpainting-Beta) but it will probably change the base image too.
> 
>     * Use differential diffusion with a soft/blurred mask, but the pipeline it's still a [PR](https://github.com/huggingface/diffusers/pull/9268).

@asomoza I tried the new community pipeline differential diffusion, thanks for pointing out it works quite well in my experience
(For large outpainting though, the model tends to just add a += uniform background)

**Comment by asomoza** [comment]
@Clement-Lelievre glad it worked better for you. Maybe for large outpainting you use other solutions and then do a second pass with Flux over it to finish the details.

For example a did a guide that [fffiloni](https://huggingface.co/fffiloni) used as a base to do a [cool space](https://huggingface.co/spaces/fffiloni/diffusers-image-outpaint) that does really good outpainting but since it was with SDXL and a lighting model, the final generation is not that good, you can use something like that and then do a img2img final pass over it.

Basically what I'm suggesting is that you use that technique with SDXL to pre fill the image with what you want before using Flux over it, that should prevent the uniform background.

**Comment by appleyang123** [comment]
> > Hi, we don't have any outpainting pipeline for any model, this is because outpainting can be done with the same inpainting pipeline you're referencing.
> > Outpainting is just inpainting but using just a part of the reference image to expand it.
> 
> (using diffusers @ [14a1b86](https://github.com/huggingface/diffusers/commit/14a1b86fc7de53ff1dbf803f616cbb16ad530e45))
> 
> hi @asomoza
> 
> maybe I'm doing this wrong but I'm not getting great results trying to do outpainting with the `FluxInpaintPipeline` (and either flux schnell or dev).
> 
> Here are the steps I follow:
> 
> 1. take an input image
> 2. pad it (to try to help the model I use the patchmatch algo to pad)
> 3. build mask
> 4. blur mask a little to get a smoother transition / or use feathering for better blending
> 5. infer
> 
> **Input image** (1024*1024)
> 
> <img alt="" width="200" height="200" src="https://private-user-images.githubusercontent.com/70368164/375824534-203ca3fb-9aca-4b2a-bce4-e2f1d1472e8b.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mjk3NjEzNDgsIm5iZiI6MTcyOTc2MTA0OCwicGF0aCI6Ii83MDM2ODE2NC8zNzU4MjQ1MzQtMjAzY2EzZmItOWFjYS00YjJhLWJjZTQtZTJmMWQxNDcyZThiLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDEwMjQlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMDI0VDA5MTA0OFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTNhMjcwNDAzYjZhMWYxNjdjMzFlYjFkODUyNzczNjc4YzQ3ZWE1ZGQ5OTM2Y2I0N2UxMzViMzFhNzk3YTkxMGQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.2_qj72vs3llbpSEEGJ0jiQlgaCpcGtctlK1Yz0IEfG8">
> **Input image + patch-match padding** (1296*1360)
> 
> <img alt="" width="200" height="200" src="https://private-user-images.githubusercontent.com/70368164/375824246-9dad70e2-867b-426a-b2bf-d4cc57e914d7.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mjk3NjEzNDgsIm5iZiI6MTcyOTc2MTA0OCwicGF0aCI6Ii83MDM2ODE2NC8zNzU4MjQyNDYtOWRhZDcwZTItODY3Yi00MjZhLWIyYmYtZDRjYzU3ZTkxNGQ3LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDEwMjQlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMDI0VDA5MTA0OFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTNjN2ExY2NkNDRiNTBmZjBlNDNjN2RhODVmNjRmNTM2NmRkNDY0NjBlMDkxZTNiMThiNzU4YTY2YTZjMzFiMDYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.RBJ7KMFhVcqASxnOFz0Muct3TsfSWTleRmXTWQj_NvQ">
> **(Blurred) Mask** (1296*1360)
> 
> <img alt="" width="200" height="200" src="https://private-user-images.githubusercontent.com/70368164/375824977-3c74e409-2cbe-4c57-be70-39e8c2ee1844.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mjk3NjEzNDgsIm5iZiI6MTcyOTc2MTA0OCwicGF0aCI6Ii83MDM2ODE2NC8zNzU4MjQ5NzctM2M3NGU0MDktMmNiZS00YzU3LWJlNzAtMzllOGMyZWUxODQ0LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDEwMjQlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMDI0VDA5MTA0OFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTY4OGJlNDM5YjZjYmRmMzViNTk2OWExMjQwYzI0ODk1NDdlOTk2M2IxNGFjZTg1OTJkOGVhMzliOTUwOGE5NmUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.k2_xW0gNUlCGZtBiVLdZOEve-go_0lBxc9gGVP3x9n8">
> **Inference (prompt:"sky and stars", 28 steps, flux-dev, guidance 3, strength 0.8)** (1296*1360)
> 
> <img alt="" width="200" height="200" src="https://private-user-images.githubusercontent.com/70368164/375825409-7d4f0c1b-9b96-4828-adcf-c70988bf02df.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mjk3NjEzNDgsIm5iZiI6MTcyOTc2MTA0OCwicGF0aCI6Ii83MDM2ODE2NC8zNzU4MjU0MDktN2Q0ZjBjMWItOWI5Ni00ODI4LWFkY2YtYzcwOTg4YmYwMmRmLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDEwMjQlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMDI0VDA5MTA0OFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTcyZGY2MGRhZGY0Yzc5YTZhNDU4OTlkNzQ0ZDA2NmYwMTkzZjllYmEyOWU5YjExNDc3ZTZmZGRiZTcxZDY0ODQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.rsRI0G5RhqGeRQfSqEirK9qRkAmMwcHdIrRFvQYMVlY">
> In the result above, while there is some degree of prompt fidelity, the transition is awful.
> 
> What am I doing wrong?
> 
> thanks

Hello, how do you blur the mask, can you paste some codes here?

**Comment by rubi-du** [comment]
@Clement-Lelievre Hi, I want to outpainting with flux too, but the result was very bad, the code as follow:
`pipe = FluxDifferentialImg2ImgPipeline.from_pretrained(
            flux_dir,
            transformer=transformer,
            controlnet=controlnet,
            text_encoder_2=text_encoder_2,
            torch_dtype=torch.bfloat16
        )
pipeline(
                prompt=prompt,
                image=image,
                mask_image=mask,
                strength=0.9,
                num_inference_steps=28,
            )`

The image is:
![background](https://github.com/user-attachments/assets/dbaad5b1-3708-4dd5-8fa8-1023fb4ec89a)
The mask is:
![mask1](https://github.com/user-attachments/assets/f4f1cd05-7a21-4272-8ebf-d86fd6f484d3)

What am I doing wrong?

thanks

**Comment by Clement-Lelievre** [comment]
> @Clement-Lelievre Hi, I want to outpainting with flux too, but the result was very bad, the code as follow: `pipe = FluxDifferentialImg2ImgPipeline.from_pretrained( flux_dir, transformer=transformer, controlnet=controlnet, text_encoder_2=text_encoder_2, torch_dtype=torch.bfloat16 ) pipeline( prompt=prompt, image=image, mask_image=mask, strength=0.9, num_inference_steps=28, )`
> 
> The image is: ![background](https://private-user-images.githubusercontent.com/154807421/389452368-dbaad5b1-3708-4dd5-8fa8-1023fb4ec89a.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzI1MjYwOTcsIm5iZiI6MTczMjUyNTc5NywicGF0aCI6Ii8xNTQ4MDc0MjEvMzg5NDUyMzY4LWRiYWFkNWIxLTM3MDgtNGRkNS04ZmE4LTEwMjNmYjRlYzg5YS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQxMTI1JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MTEyNVQwOTA5NTdaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT02ZTQ4MThlYzY1ZDdlMTVkZjgzZGEyODg1NTkxMTk2YWFiNTZmNTgyZWVhZDE2Mzc4M2JiZWU3NWI4MTg2ZTA3JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.TjD0RBuQ_vOcLurX61rc8NQU0Jr57UkmE46_t7rX1NQ) The mask is: ![mask1](https://private-user-images.githubusercontent.com/154807421/389452261-f4f1cd05-7a21-4272-8ebf-d86fd6f484d3.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzI1MjYwOTcsIm5iZiI6MTczMjUyNTc5NywicGF0aCI6Ii8xNTQ4MDc0MjEvMzg5NDUyMjYxLWY0ZjFjZDA1LTdhMjEtNDI3Mi04ZWJmLWQ4NmZkNmY0ODRkMy5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQxMTI1JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MTEyNVQwOTA5NTdaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT01MzQ1YjQ0MjQzYzBiNWU2YjZkMzhhM2U2YjM5MGFhNjM4YWU0NDg4NWJmNjVlNTZlMWVkZWY1ZTYyMWYwZTBiJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.CrUy6q_X-ecBX5oMRvzjzw-4i5TL2gkzsIlauU0sgv8) But the result is: ![flux_controlnet_inpaint (11)](https://private-user-images.githubusercontent.com/154807421/389453083-63351a19-47be-4d9f-8eef-ba82b5065acf.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzI1MjYwOTcsIm5iZiI6MTczMjUyNTc5NywicGF0aCI6Ii8xNTQ4MDc0MjEvMzg5NDUzMDgzLTYzMzUxYTE5LTQ3YmUtNGQ5Zi04ZWVmLWJhODJiNTA2NWFjZi5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQxMTI1JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MTEyNVQwOTA5NTdaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1kMzIyOWNmMzVlYTIwNGFjYmE2NzJmYmYxODhhODYwZTNiMmNkZWU5ZThhNzRjYTMzMmNlYjIyZDgxNTY0NTI2JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.7zfQbLNgKNorpn4BHEKF2B_ehgUTceSNzDhAspLT2WM)
> 
> What am I doing wrong?
> 
> thanks

hey,
have a look at [this app](https://huggingface.co/spaces/multimodalart/flux-outpainting) maybe this can inspire you

**Comment by rubi-du** [comment]
> now you can just use flux fill which was trained to do this, I tested it and it does a decent outpainting.

Oh Thanks, so exciting, I found the api doc, but only has inpainting demo, Have any demos related to outpainting？

**Comment by asomoza** [comment]
I can give you one, here's the images and result:

|source|mask|result|
|---|---|---|
|![peer_out](https://github.com/user-attachments/assets/c3555202-2c25-429b-8d07-b922b9f6ae01)|![peer_mask](https://github.com/user-attachments/assets/95556c51-5eea-46b5-abec-f67e8bae6996)|![outpaint](https://github.com/user-attachments/assets/d7cd9759-6f7f-4a43-86ef-88342626c66d)|

Here's the code:

```python
import torch

from diffusers import FluxFillPipeline
from diffusers.utils import load_image


image = load_image("https://huggingface.co/datasets/OzzyGT/testing-resources/resolve/main/flux_fill/peer_out.png")
mask = load_image("https://huggingface.co/datasets/OzzyGT/testing-resources/resolve/main/flux_fill/peer_mask.png")

repo_id = "black-forest-labs/FLUX.1-Fill-dev"


pipe = FluxFillPipeline.from_pretrained(
    repo_id, torch_dtype=torch.bfloat16,
).to("cuda")

image = pipe(
    prompt="",
    image=image,
    mask_image=mask,
    guidance_scale=3.5,
    num_inference_steps=50,
    generator=torch.Generator("cpu").manual_seed(43),
).images[0]

image.save("outpaint.png")
```

**Comment by rubi-du** [comment]
> ```python
> guidance_scale=3.5,
> ```

@asomoza Woo, Thank you so much, it works all right

**Comment by xhinker** [comment]
@asomoza Your solution works, thank you. here is my result:
input image: 
![image](https://github.com/user-attachments/assets/260ac84a-f84f-4d8a-b4b2-f741511f2207)

outpaint mask:
![image](https://github.com/user-attachments/assets/255a87f8-4ecc-4715-aa7d-f689e565d8db)

Result:
![image](https://github.com/user-attachments/assets/9676da7d-ae87-4775-90c8-152d5dc65cf0)

Code: https://github.com/xhinker/az_samples/blob/main/diffusion_models/flux/run_flux_fill_outpaint.py

**Comment by babyta** [comment]
FluxFillPipeline  I used this workflow to change the background, but it didn't work well.
